### PR TITLE
refactor(examples): HepMCWriter and Reader as algorithm

### PR DESCRIPTION
**Pull request by FabianKlimpel**
This PR modifies the `HepMCWriter` and `HepMCReader` such that they can be used as algorithms in the `Sequencer`.

### CODE REVIEW

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp [hunk 1]
```diff
@@ -1,21 +1,31 @@
...
 
 namespace ActsExamples {
 
 /// HepMC3 event reader.
-struct HepMC3ReaderAscii {
- public:
+struct HepMC3ReaderAscii : public IReader {
+  struct Config {
+    // The input directory
+    std::string inputDir = "";
```
Comment (msmk0) [inline Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp:22]:
```suggestion
    std::string inputDir;
```
Nitpicking: default initialization already provides an empty string.

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp [hunk 2]
```diff
@@ -1,21 +1,31 @@
...
+    // The stem of input file names
+    std::string inputStemFileName = "";
```
Comment (msmk0) [inline Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp:24]:
```suggestion
    std::string inputStem;
```
Stem is sufficient and consistent with e.g. the csv reader/writer.

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp [hunk 3]
```diff
@@ -1,21 +1,31 @@
...
+    // The output collection
+    std::string outputCollection = "";
```
Comment (msmk0) [inline Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp:26]: Can you name this a bit more specific. Maybe `outputEvent` similar to the existing names collection names like `outputSimHits` or `outputParticles`.

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp [hunk 4]
```diff
@@ -1,33 +1,51 @@
...
-  /// @return boolean indicator if the writing was successful
-  /// @note HepMC3 does not state a success or failure. The returned argument is
-  /// always true.
-  bool writeEvent(HepMC3::WriterAscii& writer,
-                  std::shared_ptr<HepMC3::GenEvent> event);
+struct HepMC3WriterAscii final
+    : public WriterT<std::vector<std::shared_ptr<HepMC3::GenEvent>>> {
+  struct Config {
+    // The output directory
+    std::string outputDir = "";
```
Comment (msmk0) [inline Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp:24]:
```suggestion
    std::string outputDir;
```
Nitpicking: see above.

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp [hunk 5]
```diff
@@ -1,33 +1,51 @@
...
+    // The stem of output file names
+    std::string outputStemFileName = "";
```
Comment (msmk0) [inline Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp:26]:
```suggestion
    std::string outputStem;
```
Stem is already sufficient. This is also consistent with the other configurations, e.g. the csv reader/writer.

File: Examples/Io/HepMC3/src/HepMC3Reader.cpp [hunk 6]
```diff
@@ -17,3 +20,47 @@ bool ActsExamples::HepMC3ReaderAscii::readEvent(
...
+
+  while (reader.read_event(event)) {
+    events.push_back(std::make_shared<HepMC3::GenEvent>(event));
+    event.clear();
+  }
+
+  ctx.eventStore.add(m_cfg.outputCollection, std::move(events));
+
+  return reader.failed() ? ActsExamples::ProcessCode::ABORT
+                         : ActsExamples::ProcessCode::SUCCESS;
```
Comment (msmk0) [inline Examples/Io/HepMC3/src/HepMC3Reader.cpp:65]: If the reading has a failure the output collection might be missing something. The failure check should therefore come before the collection is written to the whiteboard.

File: Examples/Io/HepMC3/src/HepMC3Writer.cpp [hunk 7]
```diff
@@ -1,20 +1,33 @@
...
+  auto path = perEventFilepath(
+      m_cfg.outputDir, m_cfg.outputStemFileName + ".hepmc3", ctx.eventNumber);
+  std::cout << "Path: " << path << std::endl;
+  HepMC3::WriterAscii writer(path);
+
+  for (const auto& event : events)
+    writer.write_event(*event);
+
+  writer.close();
+  return writer.failed() ? ActsExamples::ProcessCode::ABORT
```
Comment (msmk0) [inline Examples/Io/HepMC3/src/HepMC3Writer.cpp:31]: The failure check should occur after each write. I do not not whether i/o error persist in the writer, but it is more robust if you detect errors as soon as they can occur.

**Review by msmk0 — CHANGES_REQUESTED** [review]
@FabianKlimpel Thank you splitting this up. I added some specific comments in the code. I also have a few general questions/suggestions:

- Please add a separate header file that provides a typedef for the read/written event type. Something like `HepMC3Event` is much easier to read compared to `std::vector<std::shared_ptr<HepMC3::GenEvent>>` and it simplifies life for users. Similar to the various EDM data containers e.g. `SimHitsContainer`.
- Is is necessary to store the `HepMC3::GenEvent` by shared pointer? Would it be possible to store it by value instead, i.e. `std::vector<HepMC3::GenEvent>`? Since the collection is moved to the whiteboard and can not be modified, there is no problem of shared ownership. Its lifetime is fixed until the event processing is done anyways. Is this a requirement from HepMC?
- The reader still contain some methods (`readEvent` and `status`) that are not needed anymore to satisfy the algorithm interface. They should be removed.
- For consistency with the other I/O algorithms, I would suggest to rename the algorithms to `AsciiHepMC3{Reader,Writer}` or `HepMC3Ascii{Reader,Writer}`.
- The HepMC3 example contains both the new sequencer-based code as well as the manual reading using the HepMC3 directly. My suggestion would be to use only the sequencer-based code and remove the rest.

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp [hunk 8]
```diff
@@ -1,21 +1,31 @@
...
+// Copyright (C) 2018-2020 CERN for the benefit of the Acts project
 //
 // This Source Code Form is subject to the terms of the Mozilla Public
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 #pragma once
 
+#include "ActsExamples/Framework/IReader.hpp"
+#include <Acts/Utilities/Logger.hpp>
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp]: done

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp [hunk 9]
```diff
@@ -1,21 +1,31 @@
...
 #include <HepMC3/GenEvent.h>
 #include <HepMC3/ReaderAscii.h>
 
 namespace ActsExamples {
 
 /// HepMC3 event reader.
-struct HepMC3ReaderAscii {
- public:
+struct HepMC3ReaderAscii : public IReader {
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp]: done

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp [hunk 10]
```diff
@@ -1,33 +1,51 @@
...
- public:
-  /// @brief Writes an event to file
-  /// @param writer writer of run files
-  /// @param event storage of the event
-  /// @return boolean indicator if the writing was successful
-  /// @note HepMC3 does not state a success or failure. The returned argument is
-  /// always true.
-  bool writeEvent(HepMC3::WriterAscii& writer,
-                  std::shared_ptr<HepMC3::GenEvent> event);
+struct HepMC3WriterAscii final
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp]: done

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp (hunk 1 above)
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp]: done

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp (hunk 2 above)
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp]: Consistent yes, but not sufficient imo. Changed

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp (hunk 3 above)
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp]: outputEvents it is

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp (hunk 4 above)
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp]: done

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp (hunk 5 above)
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp]: done

File: Examples/Io/HepMC3/src/HepMC3Reader.cpp (hunk 6 above)
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/src/HepMC3Reader.cpp]: fixed

File: Examples/Io/HepMC3/src/HepMC3Writer.cpp (hunk 7 above)
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/src/HepMC3Writer.cpp]: fixed

File: Examples/Io/HepMC3/src/HepMC3Writer.cpp [hunk 11]
```diff
@@ -1,20 +1,33 @@
...
 }
 
-bool ActsExamples::HepMC3WriterAscii::status(HepMC3::WriterAscii& writer) {
-  return writer.failed();
+ActsExamples::ProcessCode ActsExamples::HepMC3WriterAscii::writeT(
+    const ActsExamples::AlgorithmContext& ctx,
+    const std::vector<std::shared_ptr<HepMC3::GenEvent>>& events) {
+  auto path = perEventFilepath(
+      m_cfg.outputDir, m_cfg.outputStemFileName + ".hepmc3", ctx.eventNumber);
+  std::cout << "Path: " << path << std::endl;
```
[the remark this answers is not in the document]
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/src/HepMC3Writer.cpp]: fixed

**Comment by FabianKlimpel** [comment]
> @FabianKlimpel Thank you splitting this up. I added some specific comments in the code. I also have a few general questions/suggestions:
> 
> * Please add a separate header file that provides a typedef for the read/written event type. Something like `HepMC3Event` is much easier to read compared to `std::vector<std::shared_ptr<HepMC3::GenEvent>>` and it simplifies life for users. Similar to the various EDM data containers e.g. `SimHitsContainer`.
> * Is is necessary to store the `HepMC3::GenEvent` by shared pointer? Would it be possible to store it by value instead, i.e. `std::vector<HepMC3::GenEvent>`? Since the collection is moved to the whiteboard and can not be modified, there is no problem of shared ownership. Its lifetime is fixed until the event processing is done anyways. Is this a requirement from HepMC?
> * The reader still contain some methods (`readEvent` and `status`) that are not needed anymore to satisfy the algorithm interface. They should be removed.
> * For consistency with the other I/O algorithms, I would suggest to rename the algorithms to `AsciiHepMC3{Reader,Writer}` or `HepMC3Ascii{Reader,Writer}`.
> * The HepMC3 example contains both the new sequencer-based code as well as the manual reading using the HepMC3 directly. My suggestion would be to use only the sequencer-based code and remove the rest.

Let me reply in a different order:
I removed `std::shared_ptr<..>` and pass the references instead. The name became shorter and I would like to avoid the typedef in another file as it has the opposite effect: It gets more frustrating to look it up. The class names were changed. The reader was just implemented due to symmetry reasons, mentioned in #516. The `HepMC3Example` is still running without the sequencer and therefore the `readEvent` and `status` are still needed for this file. This will require a clean-up of this example that should be covered by a different PR.

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp (hunk 2 above)
Comment (msmk0) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Reader.hpp]: Hmm, yes. One could argue about that. Thanks for changing it anyways to be consistent.

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Options.hpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+#include "ActsExamples/Utilities/OptionsFwd.hpp"
+
+#include "HepMC3Reader.hpp"
+#include "HepMC3Writer.hpp"
+
+namespace ActsExamples {
+
+namespace Options {
+
+/// @brief HepMC3 specific writer options
```
Comment (msmk0) [inline Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Options.hpp:20]:
```suggestion
/// Add HepMC3 specific writer options.
```
Nitpicking: `@brief` is not required.

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Options.hpp [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+void addHepMC3WriterOptions(Description& desc);
+
+/// Read the HepMC3 writer options and @return a HepMC3WriterAscii::Config
+///
+/// @param variables is the parameter map for the options
+///
+/// @returns a Config object for the HepMC3WriterAscii
+HepMC3AsciiWriter::Config readHepMC3WriterOptions(const Variables& variables);
+
+/// @brief HepMC3 specific reader options
```
Comment (msmk0) [inline Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Options.hpp:32]:
```suggestion
/// Add HepMC3 specific reader options.
```
Nitpicking: see above.

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp [hunk 14]
```diff
@@ -1,33 +1,51 @@
...
+    std::string outputStem;
+    // The input collection
+    std::string inputEvents;
+  };
 
-  /// @brief Reports the status of the writer
-  /// @param writer writer of run files
-  /// @return boolean status indicator
-  bool status(HepMC3::WriterAscii& writer);
+  /// @brief Constructor
```
Comment (msmk0) [inline Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp:31]:
```suggestion
  /// Construct the writer.
```
Nitpicking: `@brief` is not required.

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp [hunk 15]
```diff
@@ -1,33 +1,51 @@
...
+  ///
+  /// @param [in] cfg Config of the writer
+  /// @param [in] lvl The level of the logger
+  HepMC3AsciiWriter(const Config&& cfg, Acts::Logging::Level lvl);
+
+  /// @brief Writing method
```
Comment (msmk0) [inline Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp:37]:
```suggestion
  /// Writing events to file.
```
Nitpicking: see above.

**Review by msmk0 — CHANGES_REQUESTED** [review]
@FabianKlimpel thanks for adopting the fixes. This looks good and I have only some minor stylistic comments. I will leave it up to you whether you want to incorporate them or leave it as it is.

As a general comment: I would recommend not to add the HepMC3 specific input/output dir options and instead to use the general `input-dir` and `output-dir` option instead to have a consistent interface between different tools.

For my own curiosity: would it be possible to merge the multiple events read from file into a single `GenEvent`?

> I removed std::shared_ptr<..> and pass the references instead. The name became shorter and I would like to avoid the typedef in another file as it has the opposite effect: It gets more frustrating to look it up.

Ok.

> The HepMC3Example is still running without the sequencer and therefore the readEvent and status are still needed for this file. This will require a clean-up of this example that should be covered by a different PR.

Ok.

**Comment by FabianKlimpel** [comment]
> As a general comment: I would recommend not to add the HepMC3 specific input/output dir options and instead to use the general `input-dir` and `output-dir` option instead to have a consistent interface between different tools.

Is replacing the `hepmc-in/output-directory` by `in/ouput-dir` fine or is it problematic if the same tag is used multiple times?

> For my own curiosity: would it be possible to merge the multiple events read from file into a single `GenEvent`?

The event is just a collection of vertices, so this is perfectly fine.

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Options.hpp (hunk 12 above)
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Options.hpp]: done

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp (hunk 14 above)
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp]: done

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp (hunk 15 above)
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Writer.hpp]: done

File: Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Options.hpp (hunk 13 above)
Comment (FabianKlimpel) [reply Examples/Io/HepMC3/include/ActsExamples/Io/HepMC3/HepMC3Options.hpp]: done

**Comment by msmk0** [comment]
> Is replacing the `hepmc-in/output-directory` by `in/ouput-dir` fine or is it problematic if the same tag is used multiple times?

This is done almost everywhere else in the code. In that case, you would not add `input-dir` via the HepMC3 options but via the generic options. When reading the config you can check if it is set and fall-back to an empty directory if not. See how this is done e.g. for the Csv reader options.

> 
> > For my own curiosity: would it be possible to merge the multiple events read from file into a single `GenEvent`?
> 
> The event is just a collection of vertices, so this is perfectly fine.

**Review by msmk0 — APPROVED** [review]

**Pull request merged**